File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import sys, math

input = sys.stdin.readline


def init_seg(now, start, end):
    if start == end:
        tree[now] = nodes[start]
        return tree[now]
    left = now * 2
    right = now * 2 + 1
    mid = (start + end) // 2
    tree[now] = init_seg(left, start, mid) + init_seg(right, mid + 1, end)
    return tree[now]

def update_seg(target, diff, now, start, end):
    if start == end:
        tree[now] += diff
        return
    
    tree[now] += diff
    mid = (start + end) // 2
    if start <= target <= mid:
        left = now * 2
        update_seg(target, diff, left, start, mid)
    
    if mid + 1 <= target <= end:
        right = now * 2 + 1
        update_seg(target, diff, right, mid + 1, end)

def sum_seg(now, start, end, l, r):
    if r < start:
        return 0
    
    if l > end:
        return 0
    
    if start <= l <= r <= end:
        return tree[now]
    
    left = now * 2
    right = now * 2 + 1
    mid = (l + r) // 2
    _sum = sum_seg(left, start, end, l, mid) + sum_seg(right, start, end, mid + 1, r)
    return _sum
  
n, m, k = map(int, input().split())

h = math.ceil(math.log2(n))
l = 2 ** (h + 1)
nodes = []

tree = [0 for _ in range(l)]

for _ in range(n):
    node = int(input())
    nodes.append(node)

nodes.extend([0 for _ in range(2 ** h - n)])

init_seg(1, 0, len(nodes) - 1)

for _ in range(m + k):
    a, b, c = map(int, input().split())
    if a == 1:
        b -= 1
        update_seg(b, c - nodes[b], 1, 0, len(nodes) - 1)
        nodes[b] = c
    else:
        b -= 1
        c -= 1
        print(sum_seg(1, b, c, 0, len(nodes) - 1))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.rs
use std::{io::{stdin, Read, BufWriter, stdout, Write}, cmp, collections::{VecDeque, HashSet}};

fn init(tree: &mut Vec<isize>, nums: &Vec<isize>, node: usize, start: usize, end: usize) -> isize {
    if start == end {
        tree[node] = nums[start];
        return tree[node];
    }

    tree[node] = init(tree, nums, node * 2, start, (start + end) / 2) + init(tree, nums, node * 2 + 1, (start + end) / 2 + 1 , end);
    return tree[node];
}

fn sub_sum(tree: &mut Vec<isize>, node:usize, start: usize, end: usize, left: usize, right: usize) -> isize {
    if left > end || right < start {
        return 0;
    }

    if left <= start && end <= right {
        return tree[node];
    }
    return sub_sum(tree, node * 2, start, (start + end) / 2 , left, right) + sub_sum(tree, node * 2 + 1, (start + end) / 2 + 1, end, left, right);
}

fn update(tree: &mut Vec<isize>, node: usize, start: usize, end: usize, index: usize, diff: isize) {
    if index < start || index > end {
        return;
    }
    tree[node] += diff;

    if start != end {
        update(tree, node * 2, start, (start + end) / 2, index, diff);
        update(tree, node * 2 + 1, (start + end) / 2 + 1, end, index, diff);
    }
}

fn main() {
    let mut s = String::new();

    stdin().read_to_string(&mut s);

    let mut input = s.split_ascii_whitespace();
    let mut output = BufWriter::new(stdout());
    
    let n: usize = input.next().unwrap().parse().unwrap();
    let m: usize = input.next().unwrap().parse().unwrap();
    let k: usize = input.next().unwrap().parse().unwrap();

    let mut nums:Vec<isize> = vec![];
    let mut tree:Vec<isize> = vec![0; 3_000_000];

    for _ in 0..n {
        let num: isize = input.next().unwrap().parse().unwrap();
        nums.push(num);
    }
    init(&mut tree, &nums, 1, 0, n - 1);
    for _ in 0..(m + k) {
        let a: usize = input.next().unwrap().parse().unwrap();
        let b: usize = input.next().unwrap().parse().unwrap();
        let c: isize = input.next().unwrap().parse().unwrap();

        if a == 1 {
            let diff = c - nums[b - 1];
            nums[b - 1] = c;
            update(&mut tree, 1, 0, n - 1, b - 1, diff);
        }else {
            let sum = sub_sum(&mut tree, 1, 0, n - 1, b - 1, (c - 1) as usize);
            writeln!(output, "{}", sum);
        }
            
    }

}
